Doc/whatsnew/3.10.rst

```diff
@@ -2194,6 +2194,11 @@ Porting to Python 3.10
   limited API. The function is mainly useful for custom builds of Python.
   (Contributed by Petr Viktorin in :issue:`26241`)
 
+* The ``PyThreadState.use_tracing`` member has been removed to optimize Python:
+  ``tstate->use_tracing`` must be replaced with
+  ``tstate->cframe->use_tracing``.
```

Comment:
This *may* be dangerous, `state->cframe` is a stack-allocated structure so I am not sure if is valid in all accesses of `tstate->cframe` and has the same semantics. Maybe @markshannon can clarify

Reply:
Cython and greenlet now uses `tstate->cframe->use_tracing`. Do you mean that it's a bug?

Reply:
It may be

Reply:
This may be dangerous for two reasons:

1. `tstate->use_tracing` had global effect before, while `tstate->cframe->use_tracing` only affects the last executing frame. Modifiying this value only affects that particular frame. The semantics are different.
2. `tstate->cframe` points to a stack allocated variable in ceval. Accessing this from heap objects may be dangerous.

Reply:
Hum. I will design an API to abstract access to use_tracing. For now, I update my PR for the bare minimum: only document that the member is removed.

Reply:
Reading from, or writing to, `tstate->cframe->use_tracing` is OK, what is dangerous is hanging on to a pointer to `tstate->cframe`. `use_tracing` *must* always be accessed through `tstate` (and with the GIL held).

However, it may change in the future, we should make it clear that this advice only applies to 3.10.

Reply:
> Reading from, or writing to, `tstate->cframe->use_tracing` is OK,

There is also change in semantics, no? before `tstate->use_tracing` had global effect but this only affects the current frame.

Reply:
I may be wrong, but AFAIR, it was used as a flag to disable tracing _in the current frame_. That would also motivate why it was moved out of the thread state now.

Reply:
Right, but it would have effect on the whole call stack. Imagine you set it to false, then all callers would stopped tracing when they get to run again after the callee returns. iIUC now that won't happen because you are only modifying the one in the frame running and all callees

Reply:
Could setting the tracefunc to NULL be the alternative? There are some differences in 3.9 between setting use_trace to False and tracefunc to NULL, but I suspect the latter can be used. Keep in mind that this is implementation & version specific functionality so it's probably ok not to support all cases that 3.9 did.